components/supervisor/pkg/supervisor/docker.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License.AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+
+	"github.com/gitpod-io/gitpod/common-go/analytics"
+	"github.com/gitpod-io/gitpod/common-go/log"
+	"github.com/gitpod-io/gitpod/common-go/process"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/activation"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/terminal"
+)
+
+// if exit error happens one after another within dockerStartErrorBurstDuration then we're in the burst
+// and should stop trying after maxBurstDockerStartAttempts
+const (
+	maxBurstDockerStartAttempts   = 15
+	dockerStartErrorBurstDuration = time.Minute * 1
+	maxIntervalBetweenDockerStart = 15 * time.Second
+)
+
+func socketActivationForDocker(parentCtx context.Context, wg *sync.WaitGroup, term *terminal.Mux, cfg *Config, telemetry analytics.Writer, notifications *NotificationService) {
+	defer wg.Done()
+
+	ctx, cancel := context.WithCancel(parentCtx)
+	defer cancel()
+
+	burstAttempts := 0
+	backoffStrategy := backoff.NewExponentialBackOff()
+	backoffStrategy.MaxInterval = maxIntervalBetweenDockerStart
+	var lastExitErrorTime time.Time
+
+	for ctx.Err() == nil {
+		err := listenToDockerSocket(ctx, term, cfg, telemetry)
+
+		var exitError *exec.ExitError
+		if err != nil {
+			exitError, _ = err.(*exec.ExitError)
+		}
+		if exitError != nil && exitError.ExitCode() > 0 {
+			// docker-up or daemon exited with an error - we'll try again
+			// it can be a transient condition like apt is locked by another process to install prerequisites
+			// or permament like misconfigured env var, image which does not allow to install prerequisites at all
+			// or any general issue with docker daemon startup
+			if time.Since(lastExitErrorTime) <= dockerStartErrorBurstDuration {
+				// we're in the eixt error burst
+				burstAttempts++
+			} else {
+				// no burst, reset the counter
+				burstAttempts = 0
+				backoffStrategy.Reset()
+			}
+			lastExitErrorTime = time.Now()
+		} else {
+			// transient condition like a docker daemon get killed because of OOM
+			burstAttempts = 0
+			backoffStrategy.Reset()
+		}
+		nextBackOff := backoffStrategy.NextBackOff()
+
+		if nextBackOff == backoff.Stop || burstAttempts >= maxBurstDockerStartAttempts {
+			cancel()
+			// we don't want to try forever to avoid spamming our logs
+			log.WithError(err).WithField("attempts", burstAttempts).Error("cannot activate docker after maximum attempts")
+			_, _ = notifications.Notify(ctx, &api.NotifyRequest{
```

Comment:
@iQQBot not sure it is the best, the thing is that I don't expect this condition ever happen besides some automations, so maybe actually we need to show it whenever we see exit error instead? 

You also were suggesting send a notification to the client to open log file automatically?